include/oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h

```diff
@@ -105,28 +177,46 @@ class sycl_backend
         if constexpr (report_info_v<SelectionHandle, execution_info::task_completion_t> ||
                       report_value_v<SelectionHandle, execution_info::task_time_t>)
         {
-            std::chrono::steady_clock::time_point t0;
+            report_clock_type::time_point t0;
             if constexpr (report_value_v<SelectionHandle, execution_info::task_time_t>)
             {
-                t0 = std::chrono::steady_clock::now();
+                if (!is_profiling_enabled)
+                {
+                    t0 = report_clock_type::now();
+                }
             }
             auto e1 = f(q, std::forward<Args>(args)...);
-            auto e2 = q.submit([=](sycl::handler& h) {
-                h.depends_on(e1);
-                h.host_task([=]() {
-                    if constexpr (report_value_v<SelectionHandle, execution_info::task_time_t>)
-                        s.report(execution_info::task_time, (std::chrono::steady_clock::now() - t0).count());
-                    if constexpr (report_info_v<SelectionHandle, execution_info::task_completion_t>)
-                    {
+            if constexpr(report_info_v<SelectionHandle, execution_info::task_completion_t>){
+                auto e2 = q.submit([=](sycl::handler& h){
+                    h.depends_on(e1);
+                    h.host_task([=](){
                         s.report(execution_info::task_completion);
-                    }
+                    });
                 });
-            });
-            return async_waiter{e2};
+                return async_waiter{e2, std::make_shared<SelectionHandle>(s)};
```

Comment:
After changing `else if` to `if`, you have a problem with this return - it prevents the rest of the function code to execute. Therefore, supporting both task time and task completion reports still require more work.

Reply:
Fixed. Although the async_waiter is over-written since we don't know what to do when a policy both satisfies execution_info::task_time_t and execution_info::task_completion_t

Reply:
I think the following code should address that. It only creates a host task once.
```
            auto e1 = f(q, std::forward<Args>(args)...);
            async_waiter<SelectionHandle> waiter{e1,std::make_shared<SelectionHandle>(s)};

            if constexpr(report_value_v<SelectionHandle, execution_info::task_time_t>) {
                if (is_profiling_enabled)
                {
                    async_waiter_list.add_waiter(new async_waiter(waiter));
                }
            }

            bool is_host_task_needed = report_value_v<SelectionHandle, execution_info::task_time_t> && !is_profiling_enabled
                                       || report_info_v<SelectionHandle, execution_info::task_completion_t>;
            if (is_host_task_needed) {
                auto e2 = q.submit([=](sycl::handler& h){
                    h.depends_on(e1);
                    h.host_task([=](){
                        if constexpr(report_value_v<SelectionHandle, execution_info::task_time_t>){
                            if (!is_profiling_enabled) {
                                const auto tp_now = report_clock_type::now();
                                s.report(execution_info::task_time, std::chrono::duration_cast<report_duration>(tp_now - t0).count());
                            }
                        }
                        if constexpr(report_info_v<SelectionHandle, execution_info::task_completion_t>){
                            s.report(execution_info::task_completion);
                        }
                    });
                });
                waiter = async_waiter{e2, std::make_shared<SelectionHandle>(s)};
            }
            return waiter;
```

Reply:
@vossmjp I have a question though, whether in the case of a host task the returned event should point to that, or still to the submitted kernel. Given the "lazy" reporting, we cannot guarantee that waiting for the returned event ensures all associated "telemetry" is reported; but we can do that for task completion. Is it a sufficient motivation to return the host task event instead of the original kernel event?

Reply:
I think there is a tradeoff in either direction. The host task might take a while to run. If we wait for the telemetry-reporting host task to complete, this may delay the main application thread that waits on it, even though the work of the kernel is done. However, if another DS selection is made by the main thread and we do not wait for the telemetry-reporting host task, that decision is based on stale information. My preference would be to have less accurate selection instead of delaying the main thread from using a kernel result. @AnuyaWelling2801 are there measurements or tests that justify one approach over another?

Reply:
I tried waiting on the submitted kernel instead of the host task on our existing tests and they don't show much difference in timing. We should try this on existing examples to see if waiting on the existing kernel still makes it correct.

Reply:
Also, fixed the ordering